pkg/reconciler/pipelinerun/resources/validate_params.go

```diff
@@ -90,108 +88,58 @@ func ValidateObjectParamRequiredKeys(pipelineParameters []v1beta1.ParamSpec, pip
 		}
-		for j := range p.Tasks[i].Workspaces {
-			findInvalidParamArrayReference(p.Tasks[i].Workspaces[j].SubPath, arrayParams, &outofBoundParams)
+		for j := range pipelineSpec.Tasks[i].Workspaces {
+			paramsRefs = append(paramsRefs, pipelineSpec.Tasks[i].Workspaces[j].SubPath)
 		}
-		for _, wes := range p.Tasks[i].WhenExpressions {
-			findInvalidParamArrayReference(wes.Input, arrayParams, &outofBoundParams)
-			for _, v := range wes.Values {
-				findInvalidParamArrayReference(v, arrayParams, &outofBoundParams)
-			}
+		for _, wes := range pipelineSpec.Tasks[i].WhenExpressions {
+			paramsRefs = append(paramsRefs, wes.Input)
+			paramsRefs = append(paramsRefs, wes.Values...)
 		}
 	}
 
-	for i := range p.Finally {
-		findInvalidParamArrayReferences(p.Finally[i].Params, arrayParams, &outofBoundParams)
-		if p.Finally[i].IsMatrixed() {
-			findInvalidParamArrayReferences(p.Finally[i].Matrix.Params, arrayParams, &outofBoundParams)
+	for i := range pipelineSpec.Finally {
+		paramsRefs = append(paramsRefs, extractParmaRefsFromParams(pipelineSpec.Finally[i].Params)...)
+		if pipelineSpec.Finally[i].IsMatrixed() {
+			paramsRefs = append(paramsRefs, extractParmaRefsFromParams(pipelineSpec.Finally[i].Matrix.Params)...)
 		}
-		for _, wes := range p.Finally[i].WhenExpressions {
-			for _, v := range wes.Values {
-				findInvalidParamArrayReference(v, arrayParams, &outofBoundParams)
-			}
+		for _, wes := range pipelineSpec.Finally[i].WhenExpressions {
+			paramsRefs = append(paramsRefs, wes.Values...)
 		}
 	}
 
-	if outofBoundParams.Len() > 0 {
-		return fmt.Errorf("non-existent param references:%v", outofBoundParams.List())
+	// extract all array indexing references, for example []{"$(params.array-params[1])"}
+	arrayIndexParamRefs := []string{}
+	for _, p := range paramsRefs {
+		taskrun.ExtractArrayIndexingParamRefs(p, &arrayIndexParamRefs)
 	}
-	return nil
-}
-
-func extractParamIndexes(defaults []v1beta1.ParamSpec, params []v1beta1.Param) map[string]int {
-	// Collect all array params
-	arrayParams := make(map[string]int)
 
-	patterns := []string{
-		"$(params.%s)",
-		"$(params[%q])",
-		"$(params['%s'])",
-	}
-
-	// Collect array params lengths from defaults
-	for _, p := range defaults {
-		if p.Default != nil {
-			if p.Default.Type == v1beta1.ParamTypeArray {
-				for _, pattern := range patterns {
-					for i := 0; i < len(p.Default.ArrayVal); i++ {
-						arrayParams[fmt.Sprintf(pattern, p.Name)] = len(p.Default.ArrayVal)
-					}
-				}
-			}
-		}
-	}
-
-	// Collect array params lengths from pipelinerun or taskrun
-	for _, p := range params {
-		if p.Value.Type == v1beta1.ParamTypeArray {
-			for _, pattern := range patterns {
-				for i := 0; i < len(p.Value.ArrayVal); i++ {
-					arrayParams[fmt.Sprintf(pattern, p.Name)] = len(p.Value.ArrayVal)
-				}
-			}
-		}
-	}
-	return arrayParams
+	return taskrun.ValidateOutofBoundArrayParams(arrayIndexParamRefs, arrayParamsLengths)
```

Comment:
> Shouldn't taskrun params be the same as pipelinerun params
Did you mean we don't need to validate pipelinerun params? 
This function is used to validate array indexing param references are out of bound or not. I think in reconciler we don't have a pkg for params? Or how abou having 2 private functions with the same code this these 2 pkgs.

Reply:
You can feel free to create a new package. Since we need to validate that both taskrun params and pipelinerun params don't have array indexes out of bounds, shared functionality that does only this shouldn't be in the taskrun package.

Reply:
ok